### PR TITLE
Matrix-free Piola transformation for affine cells

### PR DESCRIPTION
**Pull request by NiklasWik**
This PR builds on #13591 and adds the Piola transformation for affine cells when using the matrix-free evaluation of the `FE_RaviartThomasNodal` element. It also updates the tests associated with the previous pr (with ideas from @peterrum) and adds two new ones for non-Cartesian cells.   

( #9545)

### CODE REVIEW

**Comment by peterrum** [comment]
@NiklasWik I'll take a look at this PR once the other PR is merged and you have rebased this one!

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h:5834]: This is not documented anywhere, but you do not need to compute inverse matrix for Cartesian/affine *cells*
```suggestion
          const Tensor<2, dim, VectorizedArrayType> &t_jac = this->jacobian[1];
```
This is filled in here for cells: https://github.com/dealii/dealii/blob/7e1892d3381e93ad2af8d5c2f3a51af23bef2ac1/include/deal.II/matrix_free/mapping_info.templates.h#L1212-L1217. Note that this is not done for faces (if I remember correctly) but we could do it if necessary).

File: source/dofs/dof_tools_constraints.inst.in
Comment (kronbichler) [inline source/dofs/dof_tools_constraints.inst.in:45]: I think this change in the number of instantiations and the one a few lines below should be in a separate PR. Would you mind opening that one so it can be reviewed separately?

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation.h:5834]: Ok! It does seem unnecessary to compute it twice... I'll have a look at it tomorrow (and also update the documentation)

File: source/dofs/dof_tools_constraints.inst.in
Comment (NiklasWik) [reply source/dofs/dof_tools_constraints.inst.in:45]: Yes, I agree!

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation.h:5834]: I can't find that this is used anywhere in `matrix_free/fe_evalluation.h` but I'm unsure if it is used anywhere else. Because I would like to change it to the transpose jacobian instead of just the jacobian. That way there's no need to call `transpose(invert(inv_t_jac))` if the cell is general in for instance `get_values`. I would also add it for faces.

Comment (kronbichler) [reply include/deal.II/matrix_free/fe_evaluation.h:5834]: The use is here: https://github.com/dealii/dealii/blob/c02086aecc5e6588047865e2bb4ad058bf95e346/include/deal.II/matrix_free/fe_evaluation_data.h#L1276-L1292
Regarding a call to `transpose`, I would expect that an optimizing compiler optimizes the memory operations away. (It does not always work, but I've seen it work often.) Hence, it should not matter whether you call `transpose(jac)`. Note that for general cells the Jacobians are not stored, just for the affine/Cartesian case (as the latter would be too memory intensive).

**Comment by NiklasWik** [comment]
I ended up changing to the transpose jacobian anyway because it simplifies the code a bit. Also added it to the faces. 

All matrixfree tests pass.

File: tests/matrix_free/matrix_vector_rt_face_02.cc
Comment (peterrum) [inline tests/matrix_free/matrix_vector_rt_face_02.cc:21]: You could shorten the comment here by referring to the other test and only state what is different (i.e., non-Cartesian).

File: tests/matrix_free/matrix_vector_rt_face_common.h
Comment (peterrum) [inline tests/matrix_free/matrix_vector_rt_face_common.h:64]: Thanks :)

**Review by peterrum — COMMENTED** [review]
@NiklasWik  Let me post what I have till now. I will continue with `include/deal.II/matrix_free/fe_evaluation.h` later today.

File: include/deal.II/matrix_free/mapping_info.templates.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/mapping_info.templates.h:1217]: No it's not! It has to do with the discussion above with storing the transpose or not. Not sure if this is what breaks the tests or not. I made a change in `include/deal.II/matrix_free/fe_evaluation_data.h` where this is used.

File: include/deal.II/matrix_free/mapping_info.templates.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/mapping_info.templates.h:2795]: The addition!

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation.h:5726]: I use it to get the right indicies for faces, i.e the inverse of `reorder_face_derivative_indices`

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation.h:6412]: The if-branch will be needed later when this gets implemented anyway. Or are there any other benefits other than being more nicer looking atm?

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation.h:5731]: I didn't really want to call this function more than once, and I figured that I only wanted to return a 1D array. But maybe I should do like what is done in `reorder_face_derivative_indices`, and call it for every index?

File: include/deal.II/matrix_free/mapping_info.templates.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/mapping_info.templates.h:2188]: Well it is. I wasn't sure which one of the two that was being used. Looking at it again makes it clear that I don't need this part, but only the same thing I did a couple of lines above. Since this is not used by the piola transform, and not anywhere else, I see no point in keeping it. I can keep it though, for consistency, and just remove the `reorder_face_derivative_indices` since it essentially does not do anything.

Comment (NiklasWik) [reply include/deal.II/matrix_free/mapping_info.templates.h:2188]: Or I mean it does `Assert(index < dim, ExcInternalError());` so I guess it has some value.

Comment (NiklasWik) [reply include/deal.II/matrix_free/mapping_info.templates.h:2188]: It is actually only used in 2 out of 12 places with reference-cell passed in. And what I said above is not correct. Sorry

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (kronbichler) [reply include/deal.II/matrix_free/fe_evaluation.h:5731]: I think that a static `std::array` should not be all that different from the pointer to a 1D array, as long as the compiler inlines everything correctly. I can't say whether it makes sense to call a function similar to `reorder_face_derivative_indices` for every index without benchmarking or looking at the assembly output of a compiler, as the function here is different from the one in `MappingInfo` in the sense that this one here is a hot loop, whereas the other function is "only" a setup function and thus potentially not used as often. That was also the intent for the original design by @kkormann and me, we tried to move everything expensive like shuffling around indices to the setup function to have as simple data structures in the actual operator evaluation for the regular DG face integrals.

Obviously, here we need to apply the reordering for every point in the evaluation. Or do we? Right now it is only curiosity, but do you think one could simplify the evaluators by shuffling around the normal and tangential components of the solution in such a way that the re-ordering of the Jacobian matrix on the faces would be the correct one again? I could even imagine that this might somewhat reduce the number of cases in the `evaluation_kernels` for the faces again, at least in terms of polynomial degrees that need to be expanded in normal and tangential direction. What do you think?

Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation.h:5731]: I do think you are correct! I see no immediate reason why we wouldn't be able to shuffle the components like we shuffle the derivatives in `gradients_quad` for instance, which is why we don't have to shuffle the inverse jacobian in the get/submit functions. I'll look into it.

Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation.h:5731]: Just fyi, it really made the function in `evaluation_kernels` muuuch nicer! Great idea @kronbichler, thanks! :)

Now I also (finally) see the value of having zx-coordinates  for the faces

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (peterrum) [inline include/deal.II/matrix_free/fe_evaluation.h:5744]: Personally, I would prefer if you'd move the comments inside the if-statement body.

File: include/deal.II/matrix_free/fe_evaluation_data.h
Comment (peterrum) [reply include/deal.II/matrix_free/fe_evaluation_data.h]: Does the documentation of `MappingInfo` have this information? If no could you add it there as well. Could you reference here the class/field. The documentation might get out of sync at some time.

**Review by peterrum — COMMENTED** [review]
Looks good to me! If the simplex tests should pass, what is missing in this PR? The general case can be done in a follow-up PR!?

File: include/deal.II/matrix_free/evaluation_kernels.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/evaluation_kernels.h:3103]: The order here shouldn't actually matter because I do all the offsets in `evaluate_in_face_apply`. However this makes more sense since I now always put the evaluation of the normal faces last, i.e `face_direction == component` . It also made it easier for me to write everything up on paper beforehand.

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation.h:5774]: Nope! The determinant is the same for all cases in 3D :)

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation.h:5755]: Not really. I was curious about whether or not it would be faster to use division here than to compute the determinant. @kronbichler thought it would be faster with the determinant (and I of course trust him on this) but I haven't measured it.

File: include/deal.II/matrix_free/evaluation_kernels.h
Comment (peterrum) [reply include/deal.II/matrix_free/evaluation_kernels.h:3103]: But what is `face_direction==2` in the 2D case?

Comment (NiklasWik) [reply include/deal.II/matrix_free/evaluation_kernels.h:3103]: Ah, the different calls are not with different `face_direction` but rather the `normal_dir` that I pass to the `tensor_evaluation_kernels`. Since I'm using the "general" kernels for the normal faces, this is not used for anything other than the initial offsets of the pointers.

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation.h:5750]: Oh yes, sorry for the confusion. The division was instead of previously computing the inverse of `this->jacobian[0]`. But yes, maybe the product of the diagonal elements is faster than computing the determinant. What do you think?

Comment (kronbichler) [reply include/deal.II/matrix_free/fe_evaluation.h:5750]: Yes indeed, the product of the diagonals is faster (2 multiplications in 3D) compared to the determinant, which does 4 multiplications, 5 FMAs in 3D according to https://github.com/dealii/dealii/blob/82cf0e0296706aaec7b9502b93bfb032fd385e44/include/deal.II/base/tensor.h#L2749-L2756

File: include/deal.II/matrix_free/evaluation_kernels.h
Comment (peterrum) [reply include/deal.II/matrix_free/evaluation_kernels.h:3104]: Let me reformulate the question: but what is `normal_dir==2` in the 2D case ;)

**Review by peterrum — APPROVED** [review]
Good to go from my side once you have addressed https://github.com/dealii/dealii/pull/13642/files#r862829918 and solved my confusion!

File: include/deal.II/matrix_free/evaluation_kernels.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/evaluation_kernels.h:3104]: I have to admit that the name might be misleading.. In the `tensor_evaluation_kernels` I use it to indicate in which direction the higher degree is in (which is the same as which component I'm evaluating). In `evaluate_in_face_apply`, other than passing it to the kernels, I also use it to determine the component here https://github.com/dealii/dealii/blob/a3b2bcea0c4dd7f6ef1c33bd8a196a83d72a0f8e/include/deal.II/matrix_free/evaluation_kernels.h#L3162-L3167
but also to shift the pointers here where you can see that I divide by 2 in the 2D case https://github.com/dealii/dealii/blob/a3b2bcea0c4dd7f6ef1c33bd8a196a83d72a0f8e/include/deal.II/matrix_free/evaluation_kernels.h#L3177-L3182
Passing it to the `EvalGeneral` kernels has no effect since the template parameter isn't used in the general tensor evaluator: https://github.com/dealii/dealii/blob/a3b2bcea0c4dd7f6ef1c33bd8a196a83d72a0f8e/include/deal.II/matrix_free/tensor_product_kernels.h#L201-L209
But yes, the name is a bit misleading in this function. I guess I could easily rename it to `component`(?) or something similar. `component` might not be great in this function though.. Let me know if you have any suggestions! The name kinda stuck from when I started thinking of the FE back in February.

Comment (NiklasWik) [reply include/deal.II/matrix_free/evaluation_kernels.h:3104]: Although, the name does make sense for anisotropic faces since the normal direction does not correspond to the component, e.g `component = 2`, `face_direction = 1` will have `normal_dir = 0` since we have zx-coordinates. It's just that `normal_dir = 2` does not make sense for a face in 2D as you say, but neither in 3D, but the variable is still useful for the other stuff I mentioned above. Does that help, @peterrum ?

Comment (NiklasWik) [reply include/deal.II/matrix_free/evaluation_kernels.h:3104]: (It does not make sense in 3D since the faces are in 2D)

**Comment by NiklasWik** [comment]
I'm pretty sure that it is the matrix assembly that is taking so long for the tests (on the topic of why we can't run the testcase with `dim = degree = 3`). Noticed this as I'm comparing the performance, and where the actual matrix-vector product takes a fraction of the total runtime. Still of course slow compared to the mf version.

File: include/deal.II/matrix_free/evaluation_kernels.h
Comment (peterrum) [reply include/deal.II/matrix_free/evaluation_kernels.h:3104]: What about swapping these?

```cpp
      const unsigned int face_direction = face_no / 2;

      evaluate_in_face_apply<0, EvalGeneral, EvalGeneral>(
        values_dofs,
        fe_eval,
        scratch_data,
        evaluation_flag,
        face_direction,
        subface_index,
        std::integral_constant<bool, integrate>());

      evaluate_in_face_apply<1, EvalNormal, EvalTangent>(
        values_dofs,
        fe_eval,
        scratch_data,
        evaluation_flag,
        face_direction,
        subface_index,
        std::integral_constant<bool, integrate>());

      if (dim == 3)
        evaluate_in_face_apply<2, EvalTangent, EvalNormal>(
          values_dofs,
          fe_eval,
          scratch_data,
          evaluation_flag,
          face_direction,
          subface_index,
          std::integral_constant<bool, integrate>());
```

Comment (peterrum) [reply include/deal.II/matrix_free/evaluation_kernels.h:3104]: Here

https://github.com/dealii/dealii/blob/0557c001d7c1d28860dba38ee7248aa8c65f1555/include/deal.II/matrix_free/evaluation_kernels.h#L3190-L3198

`normal_dir` can be 2 although the dimension of `EvaluatorTensorProduct` is only 1 or 2...

Comment (peterrum) [reply include/deal.II/matrix_free/evaluation_kernels.h:3104]: I see you have introduced `normal_dir` in `EvaluatorTensorProduct<evaluate_general>` during the commit https://github.com/NiklasWik/dealii/commit/6918f728e9368e9ff05c6968d60613bdea815074#diff-bd08275179f5b90beff71f4688219f973ca2cb77a58e1df1cb809aff9274fce1. I have overlooked that. Could you revert that change? There is no normal-direction in the general case. I guess the only reason for that was to be able to pass `normal_dir` to the RT type of `EvaluatorTensorProduct`. This does not feel right, since we might need in the future also something else than the normal. I would suggest to encode the information of the normal in the template arguments of the class.

Comment (NiklasWik) [reply include/deal.II/matrix_free/evaluation_kernels.h:3104]: > What about swapping these?

I prefer this order since this is the order of the faces in the quadrature points. I think that changing it only will introduce more confusion. Unless it's beneficial from a performance pov, which I doubt..?  I can add a comment explaining the order if you want. 

> Could you revert that change? 

This will require a lot of changes... And I'm not immediately sure how I would make this pretty. I would rather skip this for now, and rather address this if I have time over at the end of my thesis. But if you really think this is necessary now then I'll of course do it!

Comment (NiklasWik) [reply include/deal.II/matrix_free/evaluation_kernels.h:3104]: > normal_dir can be 2 although the dimension of EvaluatorTensorProduct is only 1 or 2...

Maybe my explanation above wasn't that good :')

**Comment by NiklasWik** [comment]
I moved the template parameter, but there are some changes in `evaluation_kernels.h` that I don't like, for instance in `interpolate_generic_raviart_thomas_apply_face(..)`. There's probably a better way to do this.

Also added a little comment with a table regarding the ordering of the faces.

File: include/deal.II/matrix_free/evaluation_kernels.h
Comment (peterrum) [inline include/deal.II/matrix_free/evaluation_kernels.h:3052]:
```suggestion
      // function. However, this is the order in which the components will be in
```

File: include/deal.II/matrix_free/evaluation_kernels.h
Comment (peterrum) [inline include/deal.II/matrix_free/evaluation_kernels.h:3059]: Thanks for adding a table! I am still not 100% happy with the code: it does not match the table. 2D does not contain any `2`s; you remove the `2` division. I would suggest to get it on master, as it is and we clean up it later on.

**Review by peterrum — APPROVED** [review]
Thanks for introducing `EvaluatorTensorProductAnisotropic`! I like it like this more! I guess by now you know me good enough that I would have preferred this change in a separate PR, since this change modifies a lot of unrelated places ;) 

@kronbichler Could you take another look at the PR and after that let's merge!

File: include/deal.II/matrix_free/evaluation_kernels.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/evaluation_kernels.h:3059]: I left out `normal_dir` for the isotropic faces in the table since the parameter is not used as `normal_dir` for those cases, but only for shifting the pointers. I'm not sure how you would like it to be instead.

File: include/deal.II/matrix_free/fe_evaluation.h
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h:6250]: Instead of copy-pasting this code, can we simply convert the `Tensor<1, dim, Tensor<1, dim, VA>>` into a `Tensor<2, dim, VA>`? Not sure if there is an automatic conversion, but otherwise you can simply copy the tensor and insert the code. It would be pity to repeat all the code here.

File: include/deal.II/matrix_free/fe_evaluation_data.h
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation_data.h:739]: What do you mean by `Only set to a useful value if on a non-Cartesian cell.`? Do you mean that the index only goes by the quadrature points `q` for the non-affine case, while for affine cells we always need to index with `[0]`? And that the affine/Cartesian case also has the actual Jacobian in the same location?

File: include/deal.II/matrix_free/fe_evaluation_data.h
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation_data.h:741]: Nit-pick:
```suggestion
   * hypercube elements, the derivatives are reorder s.t. the derivative
```

**Review by kronbichler — APPROVED** [review]
This is great work, it is clear that you put a lot of patience and effort into this work. I like the direction of this very much, we are almost there for the RT case.

File: include/deal.II/matrix_free/fe_evaluation_data.h
Comment (NiklasWik) [reply include/deal.II/matrix_free/fe_evaluation_data.h:739]: This was something that was already there and frankly something I too was confused about, but something I left untouched. I'll clearify it with that for affine/Cartesian cells only index 0 is set.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**
